packages/test/test-end-to-end-tests/src/test/gc/gcStateResetInSummaries.spec.ts

```diff
@@ -0,0 +1,276 @@
+            `Actual: ${latestSummaryContext?.referenceSequenceNumber}.`,
+        );
+        assert(latestUploadedSummary !== undefined, "Did not get a summary");
+
+        return (latestUploadedSummary.tree[channelsTreeName] as ISummaryTree)?.tree ?? latestUploadedSummary.tree;
+    }
+
+    /**
+     * Validates that GC ran by asserting that all data stores have GC state. Also, the data store whose id is
+     * unreferencedDataStoreId is marked as unreferenced as per the GC state.
+     */
+    async function validateGCRan(
+        summarizerClient: { containerRuntime: ContainerRuntime, summaryCollection: SummaryCollection },
+        unreferencedDataStoreId?: string,
+    ) {
+        const channelsTree = await getSummaryChannelsTree(summarizerClient);
+        for (const [ id, summaryObject ] of Object.entries(channelsTree)) {
+            if (summaryObject.type !== SummaryType.Tree) {
+                continue;
+            }
+
+            const gcBlob = summaryObject.tree.gc;
+            assert(gcBlob?.type === SummaryType.Blob, `DataStore ${id} should have GC blob`);
+
+            const gcSummaryDetails = JSON.parse(gcBlob.content as string) as IGarbageCollectionSummaryDetails;
+            assert(gcSummaryDetails.gcData !== undefined, `DataStore ${id} should have GC data`);
+            assert(gcSummaryDetails.usedRoutes !== undefined, `DataStore ${id} should have used routes`);
+
+            if (id === unreferencedDataStoreId) {
+                assert(summaryObject.unreferenced === true, `DataStore ${id} should be unreferenced`);
+                assert(gcSummaryDetails.unrefTimestamp !== undefined, `DataStore ${id} should have unref timestamp`);
+                assert(
+                    !gcSummaryDetails.usedRoutes.includes("") && !gcSummaryDetails.usedRoutes.includes("/"),
+                    `DataStore ${id} should not be in use`);
+            } else {
+                assert(summaryObject.unreferenced !== true, `DataStore ${id} should be referenced`);
+                assert(gcSummaryDetails.unrefTimestamp === undefined, `DataStore ${id} shouldn't have unref timestamp`);
+                assert(
+                    gcSummaryDetails.usedRoutes.includes("") || gcSummaryDetails.usedRoutes.includes("/"),
+                    `DataStore ${id} should be in use`,
+                );
+            }
+        }
+    }
+
+    /**
+     * Validates that GC did not run by asserting that no data store has GC state. They should only have used routes in
+     * GC blob that contains self route.
+     */
+    async function validateGCDidNotRun(
+        summarizerClient: { containerRuntime: ContainerRuntime, summaryCollection: SummaryCollection },
+    ) {
+        const channelsTree = await getSummaryChannelsTree(summarizerClient);
+        for (const [ id, summaryObject ] of Object.entries(channelsTree)) {
+            if (summaryObject.type !== SummaryType.Tree) {
```

Comment:
#nit `getSummaryChannelsTree` could filter out objects that aren't of this type so that clients shouldn't all do it.

Reply:
Not sure what you mean. `getSummaryChannelsTree` is a local function here. Also, that would mean we have to iterate through the entries twice and it might be better to avoid that.

Reply:
Yeah, your second point is what I mean. Probably some sort of generator function would work around that but now that I think about it I don't believe it's worth the investment. It's good as is.